src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
+                            {
+                                // it's expected if the entity is empty, try next time
+                                continue;
+                            }
+                        }
+
+                        IList<Message> messages = messages = await receiver.ReceiveAsync(_serviceBusOptions.BatchOptions.MaxMessageCount, _serviceBusOptions.BatchOptions.OperationTimeout);
+
+                        if (messages != null)
+                        {
+                            Message[] messagesArray = messages.ToArray();
+                            ServiceBusTriggerInput input = new ServiceBusTriggerInput()
+                            {
+                                Messages = messagesArray
+                            };
+                            if (_isSessionsEnabled)
+                            {
+                                input.MessageSession = receiver as IMessageSession;
+                            }
+                            else
+                            {
+                                input.MessageReceiver = receiver as MessageReceiver;
+                            }
+
+                            FunctionResult result = await _triggerExecutor.TryExecuteAsync(GetTriggerFunctionData(messagesArray, input), cancellationToken);
+
+                            if (cancellationToken.IsCancellationRequested)
+                            {
+                                return;
+                            }
+
+                            // Complete batch of messages only if the execution was succesfull
+                            if (result.Succeeded && Receiver != null)
+                            {
+                                await receiver.CompleteAsync(messagesArray.Select(x => x.SystemProperties.LockToken));
+                            }
+                        }
+                    }
+                    catch (ObjectDisposedException)
+                    {
+                        // Ignore as we are stopping the host
+                    }
+                    catch (NullReferenceException)
+                    {
+                        // Ignore as we are stopping the host
+                    }
+                    catch (Exception ex)
+                    {
+                        // Log another exception
+                        _logger.LogError($"Exception has occurred during batch listening:: {ex.ToString()}");
+                    }
+                }
+            }, cancellationToken);
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(null);
+            }
+        }
+
         public IScaleMonitor GetMonitor()
         {
             return _scaleMonitor.Value;
-        }
-    }
-}
+        }
+
+        private TriggeredFunctionData GetTriggerFunctionData(Message message, ServiceBusTriggerInput input)

```

Comment:
Recommend just defining a ToTriggerFunctionData method on ServiceBusTriggerInput, to keep all these details encapsulated. Internally it'll know the dispatch mode and can do the right thing.